GLMakie/assets/shader/voxel.frag

```diff
@@ -0,0 +1,118 @@
+    bool _; //empty structs are not allowed
+};
+
+// Sets which shading procedures to use
+{{shading}}
+
+flat in vec3 o_normal;
+in vec3 o_uvw;
+flat in int o_side;
+in vec2 o_tex_uv;
+
+#ifdef DEBUG_RENDER_ORDER
+flat in float plane_render_idx; // debug
+#endif
+
+uniform lowp usampler3D voxel_id;
+uniform uint objectid;
+
+{{uv_map_type}} uv_map;
+{{color_map_type}} color_map;
+{{color_type}} color;
+
+vec4 debug_color(uint id) {
+    return vec4(
+        float((id & uint(225)) >> uint(5)) / 5.0,
+        float((id & uint(25)) >> uint(3)) / 3.0,
+        float((id & uint(7)) >> uint(1)) / 3.0,
+        1.0
+    );
+}
+vec4 debug_color(int id) { return debug_color(uint(id)); }
+
+// unused but compilation requires it
+vec4 get_lrbt(Nothing uv_map, int id, int side) {
+    return vec4(0,0,1,1);
+}
+vec4 get_lrbt(sampler1D uv_map, int id, int side) {
+    return texelFetch(uv_map, id-1, 0);
+}
+vec4 get_lrbt(sampler2D uv_map, int id, int side) {
+    return texelFetch(uv_map, ivec2(id-1, side), 0);
+}
+
+vec4 get_color_from_texture(sampler2D color, int id) {
+    vec4 lrbt = get_lrbt(uv_map, id, o_side);
+    // compute uv normalized to voxel
+    // TODO: float precision causes this to wrap sometimes (e.g. 5.999..7.0002)
+    vec2 voxel_uv = mod(o_tex_uv, 1.0);
+    voxel_uv = mix(lrbt.xz, lrbt.yw, voxel_uv);
+    return texture(color, voxel_uv);
+}
+
+
+vec4 get_color(Nothing color, Nothing color_map, int id) {
+    return debug_color(id);
+}
+vec4 get_color(Nothing color, sampler1D color_map, int id) {
+    return texelFetch(color_map, id-1, 0);
+}
+vec4 get_color(sampler1D color, sampler1D color_map, int id) {
+    return texelFetch(color, id-1, 0);
+}
+vec4 get_color(sampler1D color, Nothing color_map, int id) {
+    return texelFetch(color, id-1, 0);
+}
+vec4 get_color(sampler2D color, sampler1D color_map, int id) {
+    return get_color_from_texture(color, id);
+}
+vec4 get_color(sampler2D color, Nothing color_map, int id) {
+    return get_color_from_texture(color, id);
+}
+
+
+void write2framebuffer(vec4 color, uvec2 id);
+
+#ifndef NO_SHADING
+vec3 illuminate(vec3 normal, vec3 base_color);
+#endif
+
+void main()
+{
+    // grab voxel id
+    int id = int(texture(voxel_id, o_uvw).x);
+
+    // id is invisible so we simply discard
+    if (id == 0) {
+        discard;
+    }
+
+    // otherwise we draw. For now just some color...
+    vec4 voxel_color = get_color(color, color_map, id);
+
+#ifdef DEBUG_RENDER_ORDER
+    if (mod(o_side, 3) != DEBUG_RENDER_ORDER)
+        discard;
+    voxel_color = vec4(plane_render_idx, 0, 0, id == 0 ? 0.01 : 1.0);
+#endif
+
+#ifndef NO_SHADING
+    voxel_color.rgb = illuminate(o_normal, voxel_color.rgb);
+#endif
+
+    // TODO: index into 3d array
+    ivec3 size = ivec3(textureSize(voxel_id, 0).xyz);
+    ivec3 idx = ivec3(o_uvw * size);
+    int lin = 1 + idx.x + size.x * (idx.y + size.y * idx.z);
+
+    // draw
+    write2framebuffer(voxel_color, uvec2(objectid, lin));
```

Comment:
I just realized that with 32 bit integers there is a risk of overflow here. Rendering a 2048x2048x1024 chunk of voxel data eats up the whole range, while taking 4GB of vram and being otherwise no problem for me to render. (At least with random data.) This should be noted in the docs. (And also the todo here needs to be removed)

Reply:
Added a small section about this to the docs